doc: improve collaborator guide, doc fast-tracking process

joyeecheung opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

doc: reorganize collaborator guide

* Add sections about first time contributions, code reviews
  and seeking consensus, waiting for approvals, testing and CI
* Move paragraphs to more suitable sections
* Update table of contents

doc: document the fast-tracking process

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
doc

joyeecheung commented on COLLABORATOR_GUIDE.md:
```diff
@@ -80,12 +102,33 @@ elevate discussion to the TSC for resolution (see below).
+* Changes that revert commit(s) and/or fix regressions.
+
+When a pull request is deemed suitable to be fast-tracked, label it with
+`fast-track`. The pull request can be landed once more than 3 collaborators
```
The number 3 here needs more discussion

mcollina replied:
I'm 👍 on fast-tracking everything that fixed a regression

refack replied:
I would go with 2, and be more explicit:
`The PR can be landed after 2 or more collaborators (excluding the author) have approved it.`

joyeecheung commented:
cc @nodejs/tsc @nodejs/collaborators

mcollina commented on COLLABORATOR_GUIDE.md:
```diff
@@ -80,12 +102,33 @@ elevate discussion to the TSC for resolution (see below).
+can be fast-tracked and may be landed after a shorter delay:
+
+* Trivial changes, for example, those fixing minor bugs or improving
+  performance without affecting API or causing other wide-reaching impact.
```
I'm 👎 on considering improving performance as a fast-track issue. In my experience, it's the contrary.

joyeecheung replied:
This was added in https://github.com/nodejs/node/pull/16135, should we remove that one? cc @addaleax

joyeecheung replied:
Also, personally I agree with @mcollina , I don't recall seeing a performance optimization PR being fast-tracked before?

addaleax replied:
This wording was added in c52e43d97c1ee, a long, long time before my PR.

I’m okay with removing performance optimizations this as well.

gibfahn replied:
Maybe a topic for more general discussion, but at least for me the triviality of the change isn't usually a reason to fast-track. My experience is that no matter how trivial the change, there's always the possibility that someone disagrees or has nits.

For me the reason to fast-track is because it is actually beneficial, e.g.
- It's a fix we want to land in a release line
- It's a code-and-learn PR, and we want to give people a good first experience (and maybe even land on the day)
- It unbreaks CI on master

So for example if it fixes a regression but won't go into a release for a couple of weeks, what's the reason to fast-track? Same for small changes.

joyeecheung replied:
@gibfahn I agree but these are just some examples, that's why we also need a rule about "n approvals to both the PR and the fast-tracking request". If a PR should not be fast-tracked, people can object to that fast-tracking request without objecting to the PR.

refack commented on COLLABORATOR_GUIDE.md:
```diff
@@ -41,36 +53,46 @@ Collaborators or additional evidence that the issue has relevance, the
+
+### First Time Contributions
+
+Courtesy should **always** be shown to individuals submitting issues and pull
```
I really liked that it was the first sentence in this guide, just after the TOC.

joyeecheung replied:
Hmm...I think we can just move the section up because this does not seem to be about "Accepting Modifications" anyway.

joyeecheung replied:
On a second thought, I don't think this paragraph should appear before explaining that "a collaborator can manage issues and pull request", it looks a bit surprising to explain "how you should handle something" before saying "you can handle something"

gibfahn commented on COLLABORATOR_GUIDE.md:
```diff
@@ -200,6 +225,20 @@ changing error messages in any way, altering expected timing of an event (e.g.
+Note that errors thrown, along with behaviors and APIs implemented by
+dependencies of Node.js (e.g. those originating from V8) are generally not
+under the control of Node.js and therefore *are not directly subject to this
+policy*. However, care should still be taken when landing updates to
```
They're not subject to the deprecation policy, but they are still tagged `semver-major`, still need the TSC signoffs, and still don't get backported to any release lines right? Why the move up to this section?

As a side note, it seems odd that this section doesn't mention that you have to put the `semver-major` label on the PR (I know it wasn't there before). Maybe it's covered somewhere else.

joyeecheung replied:
Hmm, yes this seem to belong to the "Breaking Changes and Deprecations" section. Thanks for catching that.

joyeecheung commented:
@mcollina @addaleax @gibfahn @refack I've moved the paragraphs a bit, deleted the bug-fix and optimization examples, and changed the number of required approvals to 2 (I've taken a look into the recent fast-tracked PRs and I think 2 should be safe enough. We can always revert if someone objects later).

tniessen commented on COLLABORATOR_GUIDE.md:
```diff
@@ -80,12 +109,31 @@ elevate discussion to the TSC for resolution (see below).
+* Changes that revert commit(s) and/or fix regressions.
+
+When a pull request is deemed suitable to be fast-tracked, label it with
+`fast-track`. The pull request can be landed once more than 2 collaborators
```
I think this should be `2 or more` then. LGTM otherwise.

joyeecheung commented:
@tniessen Updated to `2 or more` collaborators and added a sentence about CI is still required for fast-track PRs (it's pretty obvious, but still should be mentioned )

joyeecheung commented:
Landed as 0a1fba02a65b0b6bb53ca2d6966eaca67265e7d6, thanks!
